File: 0078-subsets/0078-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = []
        n = len(nums)

        def dfs(start, path):
            result.append(path[:])

            for i in range(start, n):
                if nums[i] not in path:
                    path.append(nums[i])
                    dfs(i+1, path)
                    path.pop()
        
        dfs(0, [])

        return result

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    vector<vector<int>> answer;
    vector<int> path;

    vector<vector<int>> subsets(vector<int>& nums) {
        answer.push_back(vector<int>());

        for (int num: nums) {
            vector<vector<int>> newSubset;
            
            for (vector<int> subset: answer) {
                subset.push_back(num);
                newSubset.push_back(subset);
            }

            for (vector<int> subset: newSubset) {
                answer.push_back(subset);
            }
        }
        
        return answer;
    }
};